### ISSUE TITLE
Complain about impure __toString implementation

### ISSUE THREAD

**Issue by ArtemGoutsoul** [vimeo/psalm#2864]
Complain about impure __toString implementation, since it can easily be unsafe and hard to detect.

Example by @Ocramius ( https://twitter.com/Ocramius/status/1230161831672664064?s=20 ): 
```
'foo' == new class { public function __toString() : string {echo 'hi'; return 'bar'; }};
```

**Comment by ArtemGoutsoul** [comment]
https://psalm.dev/r/87f7e9d5c5

**Comment by muglug** [comment]
Better example: https://psalm.dev/r/e0ba1e4b56

**Comment by muglug** [comment]
Fixed in 618ae77
